alibi_detect/utils/pytorch/prediction.py

```diff
@@ -21,37 +23,48 @@ def predict_batch(x: Union[np.ndarray, torch.Tensor], model: Union[nn.Module, nn
         Can be specified by passing either torch.device('cuda') or torch.device('cpu').
     batch_size
         Batch size used during prediction.
+    preprocess_fn
+        Optional preprocessing function for each batch.
     dtype
         Model output type, e.g. np.float32 or torch.float32.
 
     Returns
     -------
-    Numpy array or torch tensor with model outputs.
+    Numpy array, torch tensor or tuples of those with model outputs.
     """
     if device is None:
         device = torch.device('cuda' if torch.cuda.is_available() else 'cpu')
     if isinstance(x, np.ndarray):
         x = torch.from_numpy(x)
-    n = x.size(0)
+    n = len(x)
     n_minibatch = int(np.ceil(n / batch_size))
     return_np = not isinstance(dtype, torch.dtype)
-    preds = []
+    preds = []  # type: Union[list, tuple]
     with torch.no_grad():
         for i in range(n_minibatch):
             istart, istop = i * batch_size, min((i + 1) * batch_size, n)
-            preds_tmp = model(x[istart:istop].to(device))
-            if device.type == 'cuda':
-                preds_tmp = preds_tmp.cpu()
-            if return_np:
-                preds_tmp = preds_tmp.numpy()
-            preds.append(preds_tmp)
-    if return_np:
-        return np.concatenate(preds)
-    else:
-        return torch.cat(preds)
+            x_batch = x[istart:istop]
+            if isinstance(preprocess_fn, Callable):  # type: ignore
+                x_batch = preprocess_fn(x_batch)
+            preds_tmp = model(x_batch.to(device))  # type: ignore
+            if isinstance(preds_tmp, tuple):
+                if len(preds) == 0:  # init tuple with lists to store predictions
+                    preds = tuple([] for _ in range(len(preds_tmp)))
+                for j, p in enumerate(preds_tmp):
+                    if device.type == 'cuda' and isinstance(p, torch.Tensor):
+                        p = p.cpu()
+                    preds[j].append(p if not return_np or isinstance(p, np.ndarray) else p.numpy())
+            else:
+                if device.type == 'cuda' and isinstance(preds_tmp, torch.Tensor):
+                    preds_tmp = preds_tmp.cpu()
+                preds.append(preds_tmp if not return_np or isinstance(preds_tmp, np.ndarray)  # type: ignore
+                             else preds_tmp.numpy())
```

Comment:
Probably can condense the whole thing without condiditions by doing `preds_tmp = tuple(preds_tmp)` ? Not a big deal though.

Reply:
The type of `preds` in `if` is `Tuple[List[Union[np.ndarray, torch.Tensor]]]` and in `else` `List[Union[np.ndarray, torch.Tensor]]`. So your suggestion would then be to do `preds_tmp = tuple(preds_tmp)` and then add an if statement to check if `len(preds)==1`, and if so, to just return `preds[0]` at the end?

Reply:
Yea, suggestion was to convert to `tuple` so we treat both cases in the same way. Good point on the length check as I assume we want to unpack tuples of length 1?

Reply:
Is the implicit assumption that multi-output models always return a `Tuple` instead of a `List` ?

Reply:
Yes, the assumption is that a `tuple` is returned. I believe this is common practice as I can't think of a practical example where models return lists of outputs (you always could if you really wanted of course...). By extending to list types on line 50 and not returning a tuple at the end I believe the list multi-output case would also be covered if needed.

With the length check and added `tuple(preds_tmp)` I think the difference in code length will be minimal though.

Reply:
ok then fine to leave as is. btw the `else` branch doesn't consider `List`, assumes only a single tensor is returned?

Reply:
Yes, can add a check that the right type is returned by the model (`Union[np.ndarray, torch.Tensor]`) and throw an error that other return types are not covered.